flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -825,6 +837,12 @@ protected void closeTaskManagerConnection(final ResourceID resourceID, final Exc
 			slotManager.unregisterTaskManager(workerRegistration.getInstanceID());
 
 			workerRegistration.getTaskExecutorGateway().disconnectResourceManager(cause);
+			failedContainerSoFar.getAndAdd(1);
+			if (failedContainerSoFar.intValue() >= maxFailedContainers) {
+				rejectAllPendingSlotRequests(new MaximumFailedContainersException(
```

Comment:
do we need to do 
`slotManager.unregisterTaskManager` here to clear all the registered TaskManagers as well?

Reply:
No need. If other TMs can be still alive for restart strategy to make decision about whether to fail whole job. For per job cluster, if the job failed, the cluster will terminate by itself.